### PR TITLE
[gfx1250][FlyDSL] Add Ragged-M OOB Support for PTPC FP8 GEMM

### PR DESCRIPTION
**Pull request by aoli26**
## Motivation

Enable the gfx1250 FlyDSL WMMA PTPC path for FP8 a8w8 bpreshuffle GEMM in normal dispatch and tuning, covering tuned and untuned shapes without host-side M padding.

## Technical Details

- Route gfx1250 bpreshuffle FlyDSL dispatch through the WMMA PTPC backend, with a fallback for untuned shapes.
- Support ragged M and strided A/C via runtime leading dimensions, OOB TDM descriptors, and clipped partial-tile stores.
- Integrate curated gfx1250 gemm candidates into the existing bpreshuffle tuner and add tuned CSV entries.
- Fold gfx1250 coverage into `op_tests/test_gemm_a8w8.py` and remove obsolete standalone WMMA test/tune files.

## Test Plan

`python op_tests/test_gemm_a8w8.py --no-legacy` on gfx1250.

## Test Result

All tests passed.

## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

### CODE REVIEW

File: aiter/ops/flydsl/kernels/gemm_fp8fp4_gfx1250.py [hunk 1]
```diff
@@ -57,17 +57,36 @@ def _s_prefetch_inst_burst(num_pages: int, page_bytes: int = 4096):
...
+    ``early_timeout`` / ``oob_outer_bound`` are passed to the native builder
+    only when it supports them. When ``oob_outer_bound`` is needed (ragged M)
+    but the installed flydsl lacks it, the vendored builder in ``tdm_oob`` is
+    used — it carries the OOB logic so the non-tile-aligned-M GEMM path works
+    against the older flydsl this build pins.
+    """
+    if _TDM_HAS_OOB:
+        return tdm_ops.make_tensor_descriptor_2d(
+            early_timeout=early_timeout, oob_outer_bound=oob_outer_bound, **kwargs
+        )
```
[the remark this answers is not in the document]
Comment (aoli26) [reply aiter/ops/flydsl/kernels/gemm_fp8fp4_gfx1250.py]: resolved.

File: aiter/ops/flydsl/bpreshuffle_gemm_gfx1250.py [hunk 2]
```diff
@@ -116,30 +117,38 @@ def run_preshuffle_gemm_a8_gfx1250(
...
-        sa_dev = sa
-    else:
-        a_dev = torch.zeros((padded_m, K), dtype=XQ.dtype, device=XQ.device)
-        a_dev[:M] = XQ
-        sa_dev = torch.ones((padded_m,), dtype=torch.float32, device=sa.device)
-        sa_dev[:M] = sa
-
+    # Ragged M needs no host padding: the kernel clips A/scale loads and the C
+    # store to runtime M via hardware OOB.
+    a_dev = XQ.contiguous()
```
Comment (valarLip) [inline aiter/ops/flydsl/bpreshuffle_gemm_gfx1250.py]: ....

File: aiter/ops/flydsl/bpreshuffle_gemm_gfx1250.py [hunk 3]
```diff
@@ -116,30 +117,38 @@ def run_preshuffle_gemm_a8_gfx1250(
...
     b_dev = WQ.contiguous()
 
+    # M is unused at compile time (runtime i32_m drives everything), so compile with
+    # M=0 to cache one kernel per (N, K, config) and reuse it across all M.
     exe = _compile_ptpc_gemm(
-        M=padded_m,
+        M=0,
```
Comment (valarLip) [inline aiter/ops/flydsl/bpreshuffle_gemm_gfx1250.py]: ...

File: aiter/ops/flydsl/gemm_tune/flydsl_gemm_a8w8_bpreshuffle_wmma_common.py [hunk 4]
```diff
@@ -16,16 +17,26 @@
...
+
+# Mirror the ptpc fp8 LDS layout in gemm_fp8fp4_gfx1250.compile_fp8fp4_gemm so the
+# candidate filter matches the kernel's real allocation. fp8 packs 1 byte/elem
+# (packed_tile_k == tile_k); ptpc keeps no scale pool in LDS. Keep in sync with
+# that file's LDS_PAD_A_BYTES / LDS_PAD_D_BYTES / elem_bytes_d.
+_LDS_PAD_A_BYTES = 16
+_LDS_PAD_D_BYTES = 16
+_ELEM_BYTES_D = 2  # bf16 / f16 output
+
+_TILE_M = (16, 32, 64, 256)
```
Comment (valarLip) [inline aiter/ops/flydsl/gemm_tune/flydsl_gemm_a8w8_bpreshuffle_wmma_common.py]: missing 128?

File: aiter/ops/flydsl/bpreshuffle_gemm_gfx1250.py [hunk 5]
```diff
@@ -157,35 +166,35 @@ def run_preshuffle_gemm_a8_gfx1250(
     )
 
     if accumulate_fp32:
-        # fp32 atomic-accumulation scratch (zeroed; narrowed into Out below).
-        out_buf = torch.zeros((padded_m, N), dtype=torch.float32, device=Out.device)
-    elif padded_m == M:
-        out_buf = Out.contiguous()
+        # fp32 split-k atomic-accumulation scratch (zeroed; cast into Out below).
+        out_buf = torch.zeros((M, N), dtype=torch.float32, device=Out.device)
```
Comment (valarLip) [inline aiter/ops/flydsl/bpreshuffle_gemm_gfx1250.py]: ...

File: aiter/ops/flydsl/bpreshuffle_gemm_gfx1250.py [hunk 6]
```diff
@@ -157,35 +166,35 @@ def run_preshuffle_gemm_a8_gfx1250(
...
     else:
-        out_buf = torch.empty((padded_m, N), dtype=Out.dtype, device=Out.device)
+        out_buf = Out.contiguous()
```
Comment (valarLip) [inline aiter/ops/flydsl/bpreshuffle_gemm_gfx1250.py]: ...

File: op_tests/test_gemm_a8w8_bpreshuffle_gfx1250.py [hunk 7]
```diff
@@ -149,6 +160,69 @@ def test_split_k(split_k, monkeypatch):
     assert cos > 0.9999, f"split_k={split_k} drifts from sk1 (cos={cos})"
 
 
+@pytest.mark.parametrize("split_k", [2, 4])
```
Comment (valarLip) [inline op_tests/test_gemm_a8w8_bpreshuffle_gfx1250.py]: ...

**Review by valarLip — CHANGES_REQUESTED** [review]

File: aiter/configs/a8w8_bpreshuffle_tuned_gemm.csv [hunk 8]
```diff
@@ -549,3 +549,34 @@ gfx950,256,4096,57344,8192,torch.float8_e4m3fn,flydsl,979,0,1486.3101,flydsl_bpr
 gfx950,256,8192,57344,8192,torch.float8_e4m3fn,flydsl,825,0,2916.8532,flydsl_bpreshuflle_256x128x128_F8_F8_B16_2x0x1x2x0_default,2638.66,506.16,0.0
 gfx950,256,16384,57344,8192,torch.float8_e4m3fn,flydsl,825,0,5899.997,flydsl_bpreshuflle_256x128x128_F8_F8_B16_2x0x1x2x0_default,2609.01,420.85,0.0
 gfx950,256,32768,57344,8192,torch.float8_e4m3fn,ck,33,0,12218.2137,a8w8_bpreshuffle_256x256x256x128_16x16_16x16_8x32x1_8x32x1_1x32x1x8_8x8x1_2x1_intrawave_v3,2519.71,368.0,0.0
+gfx1250,256,1,2112,7168,torch.float8_e4m3fn,flydsl,261,0,8.8799,flydsl_bpreshuffle_wmma_t16x32x256_mw1_nw2_nb4_sk1_cm1_cn1,3.41,1706.12,0.0066
```
Comment (valarLip) [inline aiter/configs/a8w8_bpreshuffle_tuned_gemm.csv]: why have error ratio even we don't enable splitk?

File: op_tests/test_gemm_a8w8_bpreshuffle_gfx1250.py
Comment (valarLip) [inline op_tests/test_gemm_a8w8_bpreshuffle_gfx1250.py:1]: why not keep using op_tests/test_gemm_a8w8.py

**Review by valarLip — CHANGES_REQUESTED** [review]

File: aiter/configs/a8w8_bpreshuffle_tuned_gemm.csv (hunk 8 above)
Comment (aoli26) [reply aiter/configs/a8w8_bpreshuffle_tuned_gemm.csv]: We've narrowed the error down to the WMMA FP8 instruction itself, likely due to internal accumulator precision. A minimal handwritten HIP reproducer bitwise matches the FlyDSL kernel, which effectively rules out the FlyDSL kernel as the source of the precision issue.

File: op_tests/test_gemm_a8w8.py [hunk 9]
```diff
@@ -370,6 +405,115 @@ def test_normal_gemm_a8w8_pertoken_quant(
...
+        "pass": bool(err <= bound),
+    }
+
+
+def test_gemm_flydsl_gfx1250_features(l_dtype):
+    """gfx1250 FlyDSL feature pass: cases the production dispatch can't reach
+    (ragged M, strided A/C, split-k, warp/cluster)."""
+    dt = l_dtype[0]
+    cases = [
+        dict(m=17, n=256, k=1024, num_buffers=2, variant="dense"),
```
Comment (valarLip) [inline op_tests/test_gemm_a8w8.py]: ....

File: op_tests/test_gemm_a8w8.py [hunk 10]
```diff
@@ -370,6 +405,115 @@ def test_normal_gemm_a8w8_pertoken_quant(
...
+    split_k=1,
+    m_warp=2,
+    n_warp=2,
+    cluster_m=1,
+    cluster_n=1,
+    variant="dense",
+):
+    """One gfx1250 FlyDSL feature case (ragged M, strided A/C, split-k, or
+    warp/cluster) via the WMMA backend."""
+    torch.manual_seed(0)
```
Comment (valarLip) [inline op_tests/test_gemm_a8w8.py]: ....

File: op_tests/test_gemm_a8w8.py [hunk 11]
```diff
@@ -99,6 +99,41 @@ def run_gemm_ck_bpreshuffle(x, weight, x_scale, w_scale, dtype=dtypes.bf16):
     return aiter.gemm_a8w8_bpreshuffle(x, weight, x_scale, w_scale, None, dtype)
 
 
+@perftest(num_iters=TEST_NUM_ITERS, num_rotate_args=1, use_cuda_event=True)
+def run_gemm_flydsl_gfx1250(
```
Comment (coderfeli) [inline op_tests/test_gemm_a8w8.py]: num_rotate_args

File: op_tests/test_gemm_a8w8.py [hunk 12]
```diff
@@ -370,6 +405,114 @@ def test_normal_gemm_a8w8_pertoken_quant(
     return df
 
 
+@benchmark()
+def test_gemm_flydsl_feature(
```
Comment (coderfeli) [inline op_tests/test_gemm_a8w8.py]: need this specific feature test here?

File: op_tests/test_gemm_a8w8.py (hunk 11 above)
Comment (aoli26) [reply op_tests/test_gemm_a8w8.py]: removed.

File: aiter/configs/a8w8_bpreshuffle_tuned_gemm.csv [hunk 13]
```diff
@@ -549,3 +549,43 @@ gfx950,256,4096,57344,8192,torch.float8_e4m3fn,flydsl,979,0,1486.3101,flydsl_bpr
...
+gfx1250,256,128,2112,7168,torch.float8_e4m3fn,flydsl,9,0,8.326,flydsl_bpreshuffle_wmma_t32x64x512_mw2_nw2_nb4_sk1_cm1_cn1,465.47,1993.39,0.0096
+gfx1250,256,256,2112,7168,torch.float8_e4m3fn,flydsl,19,0,25.7514,flydsl_bpreshuffle_wmma_t128x192x128_mw2_nw2_nb4_sk1_cm1_cn1,301.0,701.13,0.0098
+gfx1250,256,512,2112,7168,torch.float8_e4m3fn,flydsl,398,0,40.0855,flydsl_bpreshuffle_wmma_t256x192x128_mw2_nw2_nb4_sk1_cm1_cn1,386.73,523.17,0.0098
+gfx1250,256,1024,2112,7168,torch.float8_e4m3fn,flydsl,398,0,68.7538,flydsl_bpreshuffle_wmma_t256x192x128_mw2_nw2_nb4_sk1_cm1_cn1,450.95,389.86,0.0096
+gfx1250,256,2048,2112,7168,torch.float8_e4m3fn,flydsl,398,0,126.0903,flydsl_bpreshuffle_wmma_t256x192x128_mw2_nw2_nb4_sk1_cm1_cn1,491.78,305.1,0.0098
+gfx1250,256,4096,2112,7168,torch.float8_e4m3fn,flydsl,398,0,240.7634,flydsl_bpreshuffle_wmma_t256x192x128_mw2_nw2_nb4_sk1_cm1_cn1,515.1,256.69,0.0098
+gfx1250,256,8192,2112,7168,torch.float8_e4m3fn,flydsl,398,0,470.1095,flydsl_bpreshuffle_wmma_t256x192x128_mw2_nw2_nb4_sk1_cm1_cn1,527.61,230.72,0.0098
+gfx1250,256,16,7168,16384,torch.float8_e4m3fn,flydsl,500,0,31.8317,flydsl_bpreshuffle_wmma_t256x256x128_mw2_nw2_nb4_sk1_cm2_cn2,118.06,3704.86,0.0237
```
Comment (valarLip) [inline aiter/configs/a8w8_bpreshuffle_tuned_gemm.csv]: tileM256 for M16?

**Review by valarLip — CHANGES_REQUESTED** [review]

**Review by valarLip — APPROVED** [review]

**Review by coderfeli — APPROVED** [review]
